File: algorithms/backtracking/n-queens/nQueens.py
N = 8

board = [[0 for _ in range(N)] for _ in range(N)]



def NQueens():
    if solveNQueens():
        printBoard(board)
    else:
        print("No Solution")

def solveNQueens():
    return solveNQueensUtil(0)

def solveNQueensUtil(col):
    if col == N:
        return True

    for i in range(N):
        if isSafe(i, col):
            board[i][col] = 1
            if solveNQueensUtil(col + 1):
                return True
            board[i][col] = 0
    return False

def isSafe(row, col):
    for i in range(col):
        if board[row][i] == 1:
            return False
    for i, j in zip(range(row, -1, -1), range(col, -1, -1)):
        if board[i][j] == 1:
            return False
    for i, j in zip(range(row, N, 1), range(col, -1, -1)):
        if board[i][j] == 1:
            return False
    return True

def printBoard(board):
    for row in  board:
        for val in row:
            print("Q" if val == 1 else ".", end=" ")
        print()




NQueens()

File: algorithms/backtracking/n-queens/NQueens.java
public class NQueens{
    static int N = 8;
    static int[][] board ;

    public static void main(String[] args) {
        for (int i = 0; i < 20; i++){
           
            N = i;
            board = new int[N][N];

            if(solveNQueens()){
                System.out.println("Solution for N = "+ i);
                printSolution();
            }
            else{
                System.out.println("No Solution for N = "+ i);
            }
        }
        // solveNQueens();
        // printSolution();

        
    }


    public static boolean solveNQueens(){
        return solve(0);
    }

    private static boolean solve(int col){
        if(col == N){
            return true;
        }
        for (int row = 0; row < N; row++) {
            if(isSafe(row, col)){
                board[row][col] = 1;
                if(solve( col + 1)) return true;
                board[row][col] = 0;
            }
        }
        return false;
    }

    private static  boolean isSafe(int row, int col){
        for(int j = 0; j < col; j++){
            if(board[row][j] == 1)return false;
        }
        for(int i = row, j = col; i >= 0 && j >= 0; i--, j--){
            if(board[i][j] == 1)return false;
        }
        for(int i = row, j = col; i < N && j >= 0; i++, j--){
            if(board[i][j] == 1)return false;
        }


        return true;
    }

    public static void printSolution(){
        for(int i = 0; i < N; i++){
            for(int j = 0; j < N; j++){
                System.out.print(board[i][j] == 1? "Q ": ". ");
            }
            System.out.println();
        }
    }

}